"insert type signature" does not respect qualified imports

For example, inserting a type signature for the expression `test` below causes an error because `test :: Ratio Integer` is inserted instead of `test :: Real.Ratio Integer`.

```haskell
import GHC.Real ((%))
import qualified GHC.Real as Real

test = 4 % 5
```

I'm using ghcide version: 0.2.0 with GHC: 8.6.5.

Thank you Neil.

If I do the following:

1. `stack --resolver=lts-14.0 new qualified-sig`
2. `cd qualified-sig`
3. Add 

```
ghc-options:
  - -Wall
```

to `package.yaml`

4. Add the code snippet

```haskell
import GHC.Real ((%))
import qualified GHC.Real as Real

test = 4 % 5
```

to `src/Lib.hs`

5. Run `stack build`

then GHC correctly outputs the qualified name in a warning:

```
/Users/runesvendsen/tmp/qualified-sig/src/Lib.hs:9:1: warning: [-Wmissing-signatures]
    Top-level binding with no type signature:
      test :: Real.Ratio Integer
  |
9 | test = 4 % 5
  | ^^^^
```

However, opening `src/Lib.hs` in VSCode and hovering over `test`, the popup reads `Top-level binding with no type signature: test :: Ratio Integer` (incorrect/unqualified name).

The only way I can get GHC to output without a qualification is if `Ratio` is actually in scope (thus making this output correct). So I wonder if, internally in ghcide, there are different values in scope than what is specified in the source file... Ie. is ghcide adding any imports to a source file when processing it (perhaps to aid autocompletion)?

--- a/src/Development/IDE/GHC/Error.hs
+++ b/src/Development/IDE/GHC/Error.hs
@@ -56,10 +56,16 @@ diagFromText diagSource sev loc msg = (toNormalizedFilePath' $ fromMaybe noFileP
 -- | Produce a GHC-style error from a source span and a message.
 diagFromErrMsg :: T.Text -> DynFlags -> ErrMsg -> [FileDiagnostic]
 diagFromErrMsg diagSource dflags e =
-    [ diagFromText diagSource sev (errMsgSpan e) $ T.pack $ Out.showSDoc dflags $
-      ErrUtils.formatErrDoc dflags $ ErrUtils.errMsgDoc e
+    [ diagFromText diagSource sev (errMsgSpan e)
+      $ T.pack $ formatErrorWithQual dflags e
     | Just sev <- [toDSeverity $ errMsgSeverity e]]
 
+formatErrorWithQual :: DynFlags -> ErrMsg -> String
+formatErrorWithQual dflags e =
+    Out.showSDoc dflags
+    $ Out.withPprStyle (Out.mkErrStyle dflags $ errMsgContext e)
+    $ ErrUtils.formatErrDoc dflags
+    $ ErrUtils.errMsgDoc e
 
 diagFromErrMsgs :: T.Text -> DynFlags -> Bag ErrMsg -> [FileDiagnostic]
 diagFromErrMsgs diagSource dflags = concatMap (diagFromErrMsg diagSource dflags) . bagToList

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1679,6 +1679,19 @@ fillTypedHoleTests = let
   , check "replace _ with foo _"
           "_" "n" "n"
           "(foo _)" "n" "n"
+  , testSession "replace _toException with E.toException" $ do
+      let mkDoc x = T.unlines
+            [ "module Testing where"
+            , "import qualified Control.Exception as E"
+            , "ioToSome :: E.IOException -> E.SomeException"
+            , "ioToSome = " <> x ]
+      doc <- createDoc "Test.hs" "haskell" $ mkDoc "_toException"
+      _ <- waitForDiagnostics
+      actions <- getCodeActions doc (Range (Position 3 0) (Position 3 maxBound))
+      chosen <- liftIO $ pickActionWithTitle "replace _toException with E.toException" actions
+      executeCodeAction chosen
+      modifiedCode <- documentContents doc
+      liftIO $ mkDoc "E.toException" @=? modifiedCode
   ]
 
 addInstanceConstraintTests :: TestTree

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -97,7 +97,6 @@ main = do
     , rootUriTests
     , asyncTests
     , clientSettingsTest
-
     , codeActionHelperFunctionTests
     ]
 
@@ -2228,7 +2227,7 @@ addSigLensesTests :: TestTree
 addSigLensesTests = let
   missing = "{-# OPTIONS_GHC -Wmissing-signatures -Wmissing-pattern-synonym-signatures -Wunused-matches #-}"
   notMissing = "{-# OPTIONS_GHC -Wunused-matches #-}"
-  moduleH = "{-# LANGUAGE PatternSynonyms #-}\nmodule Sigs where"
+  moduleH = "{-# LANGUAGE PatternSynonyms #-}\nmodule Sigs where\nimport qualified Data.Complex as C"
   other = T.unlines ["f :: Integer -> Integer", "f x = 3"]
   before  withMissing def
     = T.unlines $ (if withMissing then (missing :) else (notMissing :)) [moduleH, def, other]
@@ -2253,6 +2252,7 @@ addSigLensesTests = let
       , sigSession enableWarnings "a >>>> b = a + b"        "(>>>>) :: Num a => a -> a -> a"
       , sigSession enableWarnings "a `haha` b = a b"        "haha :: (t1 -> t2) -> t1 -> t2"
       , sigSession enableWarnings "pattern Some a = Just a" "pattern Some :: a -> Maybe a"
+      , sigSession enableWarnings "qualifiedSigTest= C.realPart" "qualifiedSigTest :: C.Complex a -> a"
       ]
       | (title, enableWarnings) <-
         [("with warnings enabled", True)

